### PR TITLE
Add a C version of the VM event hooks

### PR DESCRIPTION
**Pull request by fsfod**
Add a C version of the VM event hooks that the jitlog will be built on top of.

### CODE REVIEW

**Comment by ltratt** [comment]
I think this is one for Javier. I've invited him to the repo; when he accepts I'll add him as a reviewer.

**Comment by ltratt** [comment]
Hello @javierguerragiraldez :) It would be great to get a review from you on this.

**Comment by ltratt** [comment]
@javierguerragiraldez Tom tells me he's done adding commits to this PR, so it's now ready for your review.

@fsfod In order that Javier can do a review, I think you need to give him some more information on the PR about a) motivation for this branch b) how to use it (roughly speaking). 

Thanks both!

**Comment by fsfod** [comment]
Its the building blocks that the JIT log will be built on top of. Based on building stuff like this before in Lua I ran into limitations on what data i wanted to access and performance issues because the Lua based event system hooks can't be JIT'ed.
I expanded most the vmevent call macros to also call the my native callback and added new events that are only sent to the native callback. I just hard coded the VM event printer to be always on for testing.

File: src/lj_state.c
Comment (javierguerragiraldez) [inline src/lj_state.c]: conflicts with declaration in `lj_vmevent_printer.h`

File: src/luajit.h
Comment (javierguerragiraldez) [inline src/luajit.h]: am i right that this `*L` must be the main state (as opposed as a thread/coroutine)?  what about the `*L` that the callback gets, is that also the main one? or would be the current coroutine triggering the event? (yes, i'm really interested in coroutine context)

Comment (fsfod) [reply src/luajit.h]: L is really only used as a means to get at the global and jit states. I could add an event for coroutine switching that the jitlog could emit a marker for.

File: src/lj_state.c
Comment (fsfod) [reply src/lj_state.c]: i just added that temporally to force the event printer to always run when a state is created to make testing simpler

**Comment by ltratt** [comment]
Hello @javierguerragiraldez -- do you have any other comments or do you think this is ready for possible merging?

**Comment by javierguerragiraldez** [comment]
I think the changes to the JIT (the new events) are very good, simple and readable.  it's the example callback the one i don't like.
i think i wrote somewhere that it's too complex to be just an example, but almost any alternative would have to reuse parts of it.

**Comment by ltratt** [comment]
OK. @fsfod do you think you can address Javier's comment?

**Comment by fsfod** [comment]
@javierguerragiraldez Most of the callbacks/event handlers were copied from the wip JIT log code.  They initially started out as something close as to what the vmprinter does but i recently started use the skeleton of vmprinter as the base for the jitlog since i cleaned up there interface and added a proper non static context

**Comment by ltratt** [comment]
@fsfod Understood, but just because previous things were a bit more complex than necessary doesn't mean we have to do the same :) Do you think you can clean things up a bit?

**Comment by fsfod** [comment]
Really the only two things to make it is simpler are removing event filtering and switch to just printing out event name and nothing else

**Comment by javierguerragiraldez** [comment]
ok, i'll do a more detailed review of the vmprinter.  just to be clear, what's the specific intent?  as a foundation to be used as is?  or extended by an embedder? an example to be replaced?

**Comment by fsfod** [comment]
First to sanity check the events are generated and second as debug layer jitlog. i was thinking maybe also creating some units based on its output.  As i was working on it i thought it be useful as standalone tool where you can't use the lua based jitdump modules that's why i added the user settable output stream.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: these two types, `VMPrintContext` and `VMEventContext` are confusing.  from the code, i see that the first one is the private data for this callback module, and the second is an extra facility to set the output file.   but from the names i'd believe that `VMEventContext` is a more generic structure (for all 'event related' stuff)

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: is C89 compatibility a requirement?  some of these would be easier to read if closer to where they're used.   also not sure if it actually helps to use those instead of the `J` fields.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: please factorize the commonality of these o these options.  i'm sure it would also help with the variables use.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: for readability:
```
if (J->pt) {
    pt = ....
    pc = ...
} else {
    pt = ....
    pc = ...
}
```

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: same as on `vmevent_tracestart`. here, it could be like:
```
fprintf(context->user.output, "STOP(%d): %s %s %d\n",
      T->traceno,
      J->cur.linktype == LJ_TRLINK_INTERP ? "Compiled exit to" : "",
      proto_chunknamestr(pt),
      getlastline(context, pt, pc));
```

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: this condition is tied to the same condition on the `pc` definition, and is the only use of that variable.  i think simply removing both variables would make the whole function much more readable.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: again, a variable with a single use and non-obvious initialization only helps to pull apart related parts of the code.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: what happened here? aren't structs available?  i think all these second-level handlers should use typed arguments and the cast happen in the main callback.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: should this be static?

**Review by javierguerragiraldez — COMMENTED** [review]
to be used as example code, it could be more readable, put all pointer casting in the main callback, avoid non-obvious variable initialization, try to keep that code close to where it's used.

to be used as-is, it should be more generic.  at least factor the `fprintf()` to a user-installable function.  personally, i wouldn't like to have file operations hooked into the JIT.  keep in mind that LuaJIT is embedded as a library in another application, and probably these hooks could run in production for long periods of time (even if the output is discarded when not actively used)

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: yes i missed that one

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: fixed replaced with a proper struct. i think i was just trying to get it working quickly and forgot about it.

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: fixed by renaming by VMEventContext to VMPrintUserContext

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: Fixed by simplifying the whole function

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: Fixed by replacing with a shared function for getting the last lua location

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: Fixed by simplifying to a single call and skipping printing redundant stuff the trace start already prints

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: I had separated it out as a variable because the meaning was not obvious and it would of made the fprintf less readable if it was written inline. I renamed the variable and added a comment

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: I had been stuck with having to use c89 compilers in the past for many years while working on LuaJIT so thats just how i normally code for LuaJIT. I assume Mike Paul writes for the some lowest common denominator in compilers as well but i don't think it matters for this file.

Comment (ltratt) [reply src/lj_vmevent_printer.c]: I don't think we need to be hamstrung by C89: we should assume C11. If a platform doesn't support that, we've got much bigger worries...

**Comment by ltratt** [comment]
@fsfod Is this ready for rereview or ...?

**Comment by fsfod** [comment]
maybe one more or two more things to finish first

**Comment by ltratt** [comment]
OK -- let us know when you're ready!

**Comment by fsfod** [comment]
@ltratt it should be ready now

**Comment by ltratt** [comment]
Thanks! @javierguerragiraldez do you want to have a look at the updates?

File: src/lj_vmevent_printer.h
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.h]: undefined use of struct `VMPrintUserContext`, the warning can be silenced by adding a partial struct declaration.  just a line like
```
struct VMPrintUserContext;
```
above this.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: please use typed pointers.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: unused, untyped `eventdata` argument.  might be better left off.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: unused, untyped `exitState` argument.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: useless `data` argument

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: line too long.

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c]: what's this?

File: src/lj_jit.h
Comment (javierguerragiraldez) [inline src/lj_jit.h]: conflicts with definition in `luajit.h`

File: src/lj_vmevent_printer.c
Comment (javierguerragiraldez) [inline src/lj_vmevent_printer.c:311]: could this be conditionally compiled?  i'd like to set something in the makefile or `luaconf.h` and avoid linking stdio in the JIT core.

Comment (fsfod) [reply src/lj_vmevent_printer.c:311]: Luajit already uses stdio its the only way the lua based vmevent system can signal an uncaught Lua error happened in a vmevent hook.

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: I was going to use it as like a reason enum but worried about the number of places in existing code i would have to modify to to feed through the reason

Comment (javierguerragiraldez) [reply src/lj_vmevent_printer.c]: could (some of) that info be pulled from `J` instead?

Comment (ltratt) [reply src/lj_vmevent_printer.c]: Sounds like we should just get rid of it and maybe add it back in another PR if needed?

File: src/lj_jit.h
Comment (fsfod) [reply src/lj_jit.h]: I can't really be including the luajit.h header here I will have to make a new header for it

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: fixed

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: fixed removed

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: trying to use info from J is just guessing. Some flush trigger reasons are not from hitting limits and trace ids can be recycled if a proto is garbage collected that has root traces

Comment (javierguerragiraldez) [reply src/lj_vmevent_printer.c]: :-(

File: src/lj_vmevent_printer.h
Comment (fsfod) [reply src/lj_vmevent_printer.h]: fixed

File: src/lj_vmevent_printer.c
Comment (ltratt) [reply src/lj_vmevent_printer.c]: So I'm a bit confused: if we need this parameter to differentiate things, shouldn't we be using it? Or is it too much for this PR? Or ...?

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: removed

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: Added a proper typed struct event data but we can't really use it much here

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: fixed

File: src/lj_jit.h
Comment (fsfod) [reply src/lj_jit.h]: fixed with a new header that the eventid enum is also moved to as well as eventdata structs

File: src/lj_vmevent_printer.c
Comment (fsfod) [reply src/lj_vmevent_printer.c]: well i added it but the pr ever grows

**Comment by fsfod** [comment]
@javierguerragiraldez i think fixed the remaining issues with eventdata and there arguments

File: src/lj_vmevent_printer.h
Comment (javierguerragiraldez) [reply src/lj_vmevent_printer.h]: almost!  now i'm getting warnings about typedef redeclaration.  the quietest that works for me is to put the `typedef` in the initial partial struct.  The latter complete is just a struct definitnion.
```
typedef struct VMPrintUserContext VMPrintUserContext;
typedef void (*luaJIT_vmprint_printer)(VMPrintUserContext *ctx, const char *msg, uint32_t msglen);

struct VMPrintUserContext {
  void *userdata;                /* Optional pointer that the user can set to arbitrary data to keep track of the context */
  luaJIT_vmprint_printer output; /* User overridable function to output the messages generated */
  VM_EVENT_CLASS filter;         /* Bit mask of events to filter out of being printed */
};
```

File: src/lj_state.c
Comment (javierguerragiraldez) [reply src/lj_state.c]: this conflict persists.  the return type is different between this and the one in `lj_vmevent_printer.h`

**Comment by javierguerragiraldez** [comment]
the code now feels really complete and very readable; i think the only worrying point is the `vmevent_printer_start()` redeclaration in `lj_state.c` vs. the one in `lj_vmevent_printer.h`.

**Comment by fsfod** [comment]
that line will be reverted before i merge this it only there  to allow testing without it you would have to make your own little test program to enable the printer

File: src/lj_vmevent_printer.h
Comment (fsfod) [reply src/lj_vmevent_printer.h]: @javierguerragiraldez  what compiler and flags are you using to get those warnings?

Comment (javierguerragiraldez) [reply src/lj_vmevent_printer.h]: actually it was flagged by the IDE's syntax checker.  gcc shows it with `-Wpedantic`.

yes, with lots of noise, but this is the only warning directly attributable to these files

Comment (fsfod) [reply src/lj_vmevent_printer.h]: fixed

**Comment by fsfod** [comment]
@ltratt @javierguerragiraldez  i fixed the last typedef issue can squash?

**Review by javierguerragiraldez — APPROVED** [review]
seems very good to me now

**Comment by ltratt** [comment]
Two minor comments: the commit "TEMP automatically enable vmevent printer on state creation " seems weird. Should that be squashed into something else?

The commit "Add VMEVENT_TRACE_FLUSH thats triggered when a full trace happends" has two typos. "thats" -> "that's". "happends" -> "happens".

**Comment by fsfod** [comment]
it will be removed once you two are ok with the last 2 new commits and commit messages. I found some issues when vmevents are disabled at compile time with LUAJIT_DISABLE_VMEVENT and the last 2 commits fix it

**Comment by ltratt** [comment]
If @javierguerragiraldez is OK with those two commits, then please go ahead and reword and squash (but only when Javier says OK to do so!).

**Comment by fsfod** [comment]
@javierguerragiraldez did you get a chance to look yet it?

**Comment by javierguerragiraldez** [comment]
yes (finally).  as a whole, i like it very much.  the only weird thing is that 'TEMP' commit.

**Comment by fsfod** [comment]
@javierguerragiraldez I just did the final rebase the temp commit should now be gone

**Comment by ltratt** [comment]
@fsfod if you rebase this against master, then i think i can merge this

**Comment by fsfod** [comment]
@ltratt its now based off master but the change should be a nop except for what branch we merge in to

**Comment by ltratt** [comment]
Agreed, I think we can merge unless you say otherwise.

**Comment by fsfod** [comment]
go for it

**Pull request merged**
